exchanges/gateio/gateio_test.go

```diff
@@ -74,838 +75,3066 @@ func TestStart(t *testing.T) {
 
-	if len(payload) != 4 {
-		t.Fatal("unexpected payload length")
+func TestOptionsPositionPushData(t *testing.T) {
+	t.Parallel()
+	if err := g.wsHandleData([]byte(optionsPositionPushDataJSON)); err != nil {
+		t.Errorf("%s websocket options position push data error: %v", g.Name, err)
 	}
 }
 
-func TestGetRecentTrades(t *testing.T) {
+func TestGenerateDefaultSubscriptions(t *testing.T) {
 	t.Parallel()
-	currencyPair, err := currency.NewPairFromString("btc_usdt")
-	if err != nil {
-		t.Fatal(err)
+	if _, err := g.GenerateDefaultSubscriptions(); err != nil {
+		t.Error(err)
 	}
-	_, err = g.GetRecentTrades(context.Background(), currencyPair, asset.Spot)
-	if err != nil {
+}
+func TestGenerateDeliveryFuturesDefaultSubscriptions(t *testing.T) {
+	t.Parallel()
+	if _, err := g.GenerateDeliveryFuturesDefaultSubscriptions(); err != nil {
+		t.Error(err)
+	}
+}
+func TestGenerateFuturesDefaultSubscriptions(t *testing.T) {
+	t.Parallel()
+	if _, err := g.GenerateFuturesDefaultSubscriptions(); err != nil {
+		t.Error(err)
+	}
+}
+func TestGenerateOptionsDefaultSubscriptions(t *testing.T) {
+	t.Parallel()
+	if _, err := g.GenerateOptionsDefaultSubscriptions(); err != nil {
+		t.Error(err)
+	}
+}
+
+func TestCreateAPIKeysOfSubAccount(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() || !canManipulateRealOrders {
+		t.Skip(credInformationNotProvidedOrManipulatingRealOrdersNotAllowed)
+	}
+	if _, err := g.CreateAPIKeysOfSubAccount(context.Background(), CreateAPIKeySubAccountParams{
+		SubAccountUserID: 12345,
+		Body: &SubAccountKey{
+			APIKeyName: "12312mnfsndfsfjsdklfjsdlkfj",
+			Permissions: []APIV4KeyPerm{
+				{
+					PermissionName: "wallet",
+					ReadOnly:       false,
+				},
+				{
+					PermissionName: "spot",
+					ReadOnly:       false,
+				},
+				{
+					PermissionName: "futures",
+					ReadOnly:       false,
+				},
+				{
+					PermissionName: "delivery",
+					ReadOnly:       false,
+				},
+				{
+					PermissionName: "earn",
+					ReadOnly:       false,
+				},
+				{
+					PermissionName: "options",
+					ReadOnly:       false,
+				},
+			},
+		},
+	}); err != nil {
 		t.Error(err)
 	}
 }
 
-func TestGetHistoricTrades(t *testing.T) {
+func TestListAllAPIKeyOfSubAccount(t *testing.T) {
 	t.Parallel()
-	currencyPair, err := currency.NewPairFromString("btc_usdt")
+	if !areTestAPIKeysSet() {
+		t.Skip(credInformationNotProvided)
+	}
+	_, err := g.GetAllAPIKeyOfSubAccount(context.Background(), 1234)
 	if err != nil {
-		t.Fatal(err)
+		t.Error(err)
 	}
-	_, err = g.GetHistoricTrades(context.Background(),
-		currencyPair, asset.Spot, time.Now().Add(-time.Minute*15), time.Now())
-	if err != nil && err != common.ErrFunctionNotSupported {
+}
+
+func TestUpdateAPIKeyOfSubAccount(t *testing.T) {
```

Comment:
gateio_test.go:3078: GateIO unsuccessful HTTP status code: 404 raw response: 404 page not found